### PR TITLE
Add sourceinfo plugin

### PR DESCRIPTION
**Pull request by Alfus**
The sourceinfo plugin provides a protoc-gen-sourceinfo cmd that generates a *.sourceinfo.binpb file, for every .proto file. Each generated file contains the associated serialized `google.protobuf.SourceInfo` value. These files can then be loaded directly or embedded in a binary. 

The following helpers are provided for go:
- sourceinfo.GlobalFiles(): A drop-in replacement for protoregistry.GlobalFiles
- sourceinfo.GlobalTypes(): A drop-in replacement for protogregistry.GlobalTypes
- sourceinfo.RegisterAll(root string): Registers all *.sourceinfo.binpb files found under the root dir.
- sourceinfo.RegisterAllFS(fsys fs.FS, root string): Same as above, except uses the given file system.

### CODE REVIEW

File: cmd/version.go [hunk 1]
```diff
@@ -12,13 +12,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package protoschema
```
Comment (jhump) [inline cmd/version.go:15]: Why move this out of internal? Is this really something that should be part of exported API? Maybe it belongs in the `internal/cmd` package instead?

File: sourceinfo.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+var (
+	// SourceInfoGlobalTypes is a replacement for protoregistry.GlobalTypes that includes
+	// all registered source info.
+	SourceInfoGlobalTypes = sourceinfo.GlobalTypes
+	// SourceInfoGlobalFiles is a replacement for protoregistry.GlobalFiles that includes
+	// all registered source info.
+	SourceInfoGlobalFiles = sourceinfo.GlobalFiles
```
Comment (jhump) [inline sourceinfo.go:35]: As declared, these leak types from the `sourceinfo` package into the exported API of this package. Also, the type of `sourceinfo.GlobalFiles` is icky (there are comments in the code; it's only that way for back-compat, since that module is post-v1.0).

How about instead:
```
SourceInfoGlobalFiles protodesc.Resolver = sourceinfo.GlobalFiles

SourceInfoGlobalTypes interface {
	protoregistry.MessageTypeResolver
	protoregistry.ExtensionTypeResolver
	RangeExtensionsByMessage(message protoreflect.FullName, f func(protoreflect.ExtensionType) bool)
} = sourceinfo.GlobalTypes
```

Also, these seem awkward to be exported API in the `protoschema` package, since they seem rather specialized and not directly related to manipulating protobuf schemas and schema formats. Maybe they should be moved into a sub-package?

File: cmd/version.go (hunk 1 above)
Comment (Alfus) [reply cmd/version.go]: I just didn't want it to use the protoschema package name. I moved it to 'internal'.

File: sourceinfo.go (hunk 2 above)
Comment (Alfus) [reply sourceinfo.go]: done

File: sourceinfo/sourceinfo.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/jhump/protoreflect/desc/sourceinfo"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+var (
+	// SourceInfoGlobalTypes is a replacement for protoregistry.GlobalTypes that includes
```
Comment (jhump) [inline sourceinfo/sourceinfo.go:33]: nit: after refactor, doc comment now refers to wrong name

**Review by jhump — APPROVED** [review]
Just one remark about fixing up doc comments, otherwise LGTM

**Review by mfridman — APPROVED** [review]
Seems fine, only question is whether we want this available as a Buf-managed plugin over on https://github.com/bufbuild/plugins?

We also have https://github.com/bufbuild/tools which is a collection of public tools/plugins we use. But this plugin does appear to be closer to what this (bufbuild/protoschema-plugins) repository aims to be.

tl;dr - is this plugin something the rest of the community would want, or something we're primarily interested in? I'm not sure.

File: sourceinfo/sourceinfo.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/jhump/protoreflect/desc/sourceinfo"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+var (
+	// GlobalTypes is a replacement for protoregistry.GlobalTypes that includes
```
Comment (mfridman) [inline sourceinfo/sourceinfo.go:33]: Nit, the GlobalFiles and GlobalTypes comments are swapped.

**Comment by Alfus** [comment]
The target use case is to enable runtime, reflection-based 'plugins'. For example, using a protobuf schema as the basis for pflags requires access to the documentation, but does not require code gen otherwise. This plugin is a building block for
any similar reflection-based used cases.

**Comment by mfridman** [comment]
Probably worth a mention in the README.md, even if just a brief description of what this plugin does?

File: internal/cmd/jsonschema-generate-testdata/main.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2023 Buf Technologies, Inc.
+// Copyright 2024 Buf Technologies, Inc.
```
Comment (bufdev) [inline internal/cmd/jsonschema-generate-testdata/main.go:1]: Copyright years should be additive, so this (and everything else) should be `2023-2024`.

File: internal/cmd/sourceinfo-generate-testdata/main.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	if len(os.Args) != 2 {
+		return fmt.Errorf("usage: %s [output dir]", os.Args[0])
+	}
+	outputDir := os.Args[1]
+	// Make sure the directory exists
+	if err := os.MkdirAll(outputDir, 0755); err != nil {
+		return err
+	}
+
+	testFiles, err := golden.GetTestFiles("./internal/testdata")
```
Comment (bufdev) [inline internal/cmd/sourceinfo-generate-testdata/main.go:48]: `filepath.FromSlash`?

File: internal/cmd/sourceinfo-generate-testdata/main.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	if err != nil {
+		return err
+	}
+	// TODO: Use normal the plugin to generate golden files
+	includePrefixes := []string{
+		filepath.FromSlash("buf/protoschema/test/"),
+		filepath.FromSlash("bufext/cel/expr/conformance/proto3/"),
+	}
+	err = nil
```
Comment (bufdev) [inline internal/cmd/sourceinfo-generate-testdata/main.go:57]: This is unnecessary - we know that the error is nil

File: internal/cmd/sourceinfo-generate-testdata/main.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	testFiles.RangeFiles(func(testDesc protoreflect.FileDescriptor) bool {
+		if !shouldIncludeFile(testDesc, includePrefixes) {
+			return true
+		}
+		fileName := pluginsourceinfo.GetSourceInfoPath(testDesc)
+		filePath := filepath.Join(outputDir, fileName)
+		// Create any missing directories
+		if err = os.MkdirAll(filepath.Dir(filePath), 0755); err != nil {
```
Comment (bufdev) [inline internal/cmd/sourceinfo-generate-testdata/main.go:65]: `err :=`, always (this `err` should be scoped to this if statement)

File: internal/cmd/sourceinfo-generate-testdata/main.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+			err = fmt.Errorf("failed to create directory for %s: %w", filePath, err)
+			return false
+		}
+		var data []byte
+		data, err = pluginsourceinfo.GenFileContents(testDesc)
+		if err != nil {
+			err = fmt.Errorf("failed to generate source info for %s: %w", testDesc.FullName(), err)
+			return false
+		}
+		if err = os.WriteFile(filePath, data, 0600); err != nil {
```
Comment (bufdev) [inline internal/cmd/sourceinfo-generate-testdata/main.go:75]: `err :=`

File: internal/protoschema/golden/golden.go [hunk 10]
```diff
@@ -23,13 +23,14 @@ import (
...
 	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
 	"google.golang.org/protobuf/types/descriptorpb"
 	"google.golang.org/protobuf/types/dynamicpb"
 )
 
 // GetTestDescriptors returns the test descriptors that were generated from the ./internal/proto
 // directory.
-func GetTestDescriptors(testdataPath string) ([]protoreflect.MessageDescriptor, error) {
+func GetTestFileDescriptorSet(testdataPath string) (*descriptorpb.FileDescriptorSet, error) {
```
Comment (bufdev) [inline internal/protoschema/golden/golden.go:33]: If you are changing the function name, the comment should change as well.

File: internal/protoschema/golden/golden.go [hunk 11]
```diff
@@ -39,9 +40,25 @@ func GetTestDescriptors(testdataPath string) ([]protoreflect.MessageDescriptor,
 	if err = (&protojson.UnmarshalOptions{DiscardUnknown: true}).Unmarshal(input, fdset); err != nil {
 		return nil, fmt.Errorf("failed to parse file descriptor set at %q: %w", inputPath, err)
 	}
+	return fdset, nil
+}
+
+func GetTestFiles(testdataPath string) (*protoregistry.Files, error) {
```
Comment (bufdev) [inline internal/protoschema/golden/golden.go:46]: Needs Godoc

File: internal/protoschema/golden/golden.go [hunk 12]
```diff
@@ -39,9 +40,25 @@ func GetTestDescriptors(testdataPath string) ([]protoreflect.MessageDescriptor,
...
+	}
 	files, err := protodesc.NewFiles(fdset)
 	if err != nil {
-		return nil, fmt.Errorf("failed to link file descriptor set at %q: %w", inputPath, err)
+		return nil, fmt.Errorf("failed to link file descriptor set at %q: %w", testdataPath, err)
+	}
+	return files, nil
+}
+
+func GetTestDescriptors(testdataPath string) ([]protoreflect.MessageDescriptor, error) {
```
Comment (bufdev) [inline internal/protoschema/golden/golden.go:58]: Needs Godoc

File: internal/protoschema/plugin/pluginsourceinfo/pluginsourceinfo.go [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/bufbuild/protoplugin"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+const FileExtension = ".sourceinfo.binpb"
```
Comment (bufdev) [inline internal/protoschema/plugin/pluginsourceinfo/pluginsourceinfo.go:28]: Needs Godoc - please ensure all of your public types have Godoc before putting up PRs here, will not comment for rest of files

File: sourceinfo/sourceinfo.go [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+var (
+	// GlobalFiles is a replacement for protoregistry.GlobalFiles that includes
+	// all registered source info.
+	GlobalFiles protodesc.Resolver = sourceinfo.GlobalFiles
```
Comment (bufdev) [inline sourceinfo/sourceinfo.go:35]: Global variables are banned - Make this API not use global variables

File: internal/cmd/sourceinfo-generate-testdata/main.go (hunk 8 above)
Comment (Alfus) [reply internal/cmd/sourceinfo-generate-testdata/main.go:65]: this is a lambda that is reporting the error to the outter scope, it cannot use :=

File: internal/cmd/sourceinfo-generate-testdata/main.go (hunk 9 above)
Comment (Alfus) [reply internal/cmd/sourceinfo-generate-testdata/main.go:75]: needs to be = because of the lambda

File: internal/cmd/sourceinfo-generate-testdata/main.go (hunk 8 above)
Comment (bufdev) [reply internal/cmd/sourceinfo-generate-testdata/main.go]: Then it shouldn't be set inside of an `if` statement - that's confusing to readers. I'm actually surprised that there isn't a linter that bans this in `golangci-lint`

File: internal/cmd/sourceinfo-generate-testdata/main.go (hunk 9 above)
Comment (bufdev) [reply internal/cmd/sourceinfo-generate-testdata/main.go]: Same

File: README.md [hunk 15]
```diff
@@ -162,6 +162,19 @@ Results in the following JSON Schema files:
 
 </details>
 
+## SourceInfo
+
+The sourceinfo plugin provides a protoc-gen-sourceinfo command that generates an `.sourceinfo.binpb` file, for every `.proto` file.
+Each generated file contains the associated serialized `google.protobuf.SourceInfo` value.
```
Comment (bufdev) [inline README.md:168]: There is no type called `google.protobuf.SourceInfo`. Do you mean `google.protobuf.SourceCodeInfo`? If so, this calls into question the naming of a lot of things here (package, plugin, etc).

Additionally, a `SourceCodeInfo` (if this is the intent) does not make a ton sense without an associated `FileDescriptorProto` - why not make the plugin provide the entire `FileDescriptorProto/FileDescriptorSet`?

File: README.md [hunk 16]
```diff
@@ -162,6 +162,19 @@ Results in the following JSON Schema files:
...
+These files can then be loaded directly or embedded in a binary.
+
+The following helpers are provided for go via `import "github.com/bufbuild/protoschema-plugins/sourceinfo"`:
+
+- `sourceinfo.GlobalFiles()`: A drop-in replacement for protoregistry.GlobalFiles
```
Comment (bufdev) [inline README.md:173]: How can this be a drop-in if you just are serializing the `google.protobuf.SourceCodeInfo` file?

File: internal/protoschema/plugin/pluginsourceinfo/pluginsourceinfo.go [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// GetSourceInfoPath returns the path to the source info file for the given file descriptor.
+func GetSourceInfoPath(fileDescriptor protoreflect.FileDescriptor) string {
+	path := fileDescriptor.Path()
+	path = strings.TrimSuffix(path, ".proto")
+	return fmt.Sprintf("%s%s", path, FileExtension)
+}
+
+// GenFileContents generates the source info file contents for the given file descriptor.
```
Comment (bufdev) [inline internal/protoschema/plugin/pluginsourceinfo/pluginsourceinfo.go:63]: I don't know what this means - what is the "source info file contents"? Can you explain in the documentation what this is? All I am getting back is a `[]byte`.

File: sourceinfo/sourceinfo.go [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/bufbuild/protoschema-plugins/internal/protoschema/plugin/pluginsourceinfo"
+	"github.com/jhump/protoreflect/desc/sourceinfo"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// GlobalFiles is a replacement for protoregistry.GlobalFiles that includes
```
Comment (bufdev) [inline sourceinfo/sourceinfo.go:32]: Needs much more extensive documentation. As a reader, I don't know what "all registered source info" is, and re: comment above, I don't know how this replaces `protoregistry.GlobalFiles` (and I should not have to read the code to find out if it actually does or how it does)

File: sourceinfo/sourceinfo.go (hunk 14 above)
Comment (bufdev) [reply sourceinfo/sourceinfo.go]: This was not resolved - you just changed this from a global variable to a static function that wraps a global variable.

**Review by bufdev — CHANGES_REQUESTED** [review]

File: sourceinfo/sourceinfo.go (hunk 14 above)
Comment (Alfus) [reply sourceinfo/sourceinfo.go]: This is how the sourceinfo package exposes the functionality. Are you suggesting that package can't be used? Or do you have another solution in mind?

File: README.md (hunk 15 above)
Comment (Alfus) [reply README.md:168]: Nice typo find! I was following @jhump's naming scheme, I could switch it to source _code_ info every where if you think that would provide more clarity.
The normal code gen already comes packed with the Descriptors for normal reflection, this plugin provides just the part that is stripped out. @jhump's sourceinfo library nicely glues everything together at runtime, they same could be done in other languages.

File: README.md (hunk 16 above)
Comment (Alfus) [reply README.md:173]: This is hooking into @jhump's library to provide a drop in replacement, backed by normal protoregistry values.
